### PR TITLE
Convert to pug templates. Fix #36

### LINKED ISSUE #36: Consider converting to pug templates
Compiling HTML and importing other HTML snippets is a bit unwieldy.
Several Roll20 examples use `pug` to template and assemble HTML snippets into character sheets.
[Pug](https://pugjs.org/language/includes.html)

Some example Roll20 Pug usage:
1. https://github.com/Roll20/roll20-character-sheets/blob/master/ADnD_2E_Revised/pug/weaponsOverview.pug
2. https://github.com/Roll20/roll20-character-sheets/blob/master/backbone/source/backbone.pug

@johnlamping thoughts?

### PR DESCRIPTION
**Pull request by jazeee**
Compiling HTML and importing other HTML snippets is a bit unwieldy.

Several Roll20 examples use `pug` to template and assemble HTML snippets into character sheets.

[Pug](https://pugjs.org/language/includes.html)

Some example Roll20 Pug usage:
1. https://github.com/Roll20/roll20-character-sheets/blob/master/ADnD_2E_Revised/pug/weaponsOverview.pug
2. https://github.com/Roll20/roll20-character-sheets/blob/master/backbone/source/backbone.pug

Add `beautify-js` to clean up HTML.

Fix #36

### CODE REVIEW

File: bin/main.html [hunk 1]
```diff
@@ -430,7 +430,7 @@
       <fieldset class="repeating_extraadvantage">
         <span name="attr_extraadvantagename" class="extraadvantagename">
         </span>
-        <fieldset>
+      </fieldset>
```
Comment (jazeee) [inline bin/main.html:441]: As part of testing out pug, I discovered this error.

File: ui/main.pug [hunk 2]
```diff
@@ -0,0 +1,25 @@
+include rollSection/rollSection.html
+= '\n'
```
Comment (jazeee) [inline ui/main.pug:2]: Blank lines are a bit of a pain in `pug`

File: ui/overviewTab/overviewTab.html [hunk 3]
```diff
@@ -71,7 +71,7 @@
       <fieldset class="repeating_extraadvantage">
         <span name="attr_extraadvantagename" class="extraadvantagename">
         </span>
-        <fieldset>
+      </fieldset>
```
Comment (jazeee) [inline ui/overviewTab/overviewTab.html]: Separated into:
https://github.com/epic-power-rpg/roll20/pull/38

File: ui/overviewTab/overviewTab.pug [hunk 4]
```diff
@@ -0,0 +1,107 @@
+// Overview section
+.overview.section
```
Comment (jazeee) [inline ui/overviewTab/overviewTab.pug:37]: This is an example of the above html converted to pug via an online tool.
I had to adjust it to get close to the same html, although there are some HTML differences.
For example, the `class` attribute will come before `name` attribute.

Some difference examples:

image

```
+    <!-- Feats-->
     <div class="feat-overview no-add-or-modify-buttons">
       <div class="table-header">
         <div class="featextraordinaryletter">Extr</div>
         <div class="featname">Epic Feat</div>
       </div>
       <fieldset class="repeating_feat">
-        <span name="attr_featextraordinaryletter" class="featextraordinaryletter"></span>
-        <span name="attr_featname" class="featname"> </span>
+        
+        <span class="featextraordinaryletter" name="attr_featextraordinaryletter"></span>
+        <span class="featname" name="attr_featname"> </span>
       </fieldset>
     </div>
```

File: ui/overviewTab/overviewTab.pug [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      fieldset.repeating_weapon
+        .weapons-overview-row-container
+          |
+          | #[span.equipmentname(name='attr_weaponname')]
+          input.equipmentcount(name='attr_weaponcount' type='number' value='1')
+          |
+          | #[span.equipmentdamage(name='attr_weapondamage')]
+          | #[span.equipmentrange(name='attr_weaponrange')]
+        |
+        | #[span.equipmentnotes(name='attr_weaponnotes')]
```
Comment (jazeee) [inline ui/overviewTab/overviewTab.pug]: I had to put these extra 
```
|
| #[span...]
```
to make `span` elements be their own line.
By default `pug` puts `inline` elements on one line, like:
```
<fieldset...><span class="equipmentnotes" name.../></fieldset>
```

Personally, I'd like three lines...
That being said, we can also not worry about html formatting so much. (or we can post-prettify it)

File: ui/main.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+const { argv } = yargs
+  .option('watch', {
+    alias: 'w',
+    description: 'Watch for pug changes',
+  })
+  .help();
+
+function renderHtml() {
+  const html = pug.renderFile(`${__dirname}/main.pug`, { pretty: '  ' });
```
Comment (jazeee) [inline ui/main.js]: `pug` has deprecated `pretty` option, but this helps keep the output HTML legible.
We could also look into post-processing to prettify the HTML.

File: ui/main.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+const path = require('path');
+const yargs = require('yargs');
+const chokidar = require('chokidar');
+
+const { argv } = yargs
+  .option('watch', {
+    alias: 'w',
+    description: 'Watch for pug changes',
+  })
+  .help();
```
Comment (jazeee) [inline ui/main.js:26]: This makes `yarn pug --help` work.
`yarn pug --watch` is now an option.

File: ui/main.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  const watcher = chokidar.watch(`${__dirname}/**/*.{html,pug}`, {
+    persistent: true
+  });
+  watcher.on('ready', renderHtml);
+  watcher.on('change', (filename) => {
+    console.log(`${filename} changed`);
+    renderHtml();
+  });
+} else {
+  renderHtml();
```
Comment (jazeee) [inline ui/main.js:50]: This is a single compile `yarn pug`

File: ui/main.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  fs.writeFileSync(outputFilename, html);
+  console.log(`Wrote ${outputFilename}`);
+}
+
+if (argv.watch) {
+  const watcher = chokidar.watch(`${__dirname}/**/*.{html,pug}`, {
+    persistent: true
+  });
+  watcher.on('ready', renderHtml);
+  watcher.on('change', (filename) => {
```
Comment (jazeee) [inline ui/main.js:45]: `render` on watcher completing scan, and then on any changes to watched files.

File: ui/overviewTab/overviewTab.pug [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+      .conditional-something
+        input(name='attr_advantage_mage_count' type='hidden')
+        div
+          | Mage
+          | #[span(name='attr_advantage_mage_count')]
+      .conditional-something
+        input(name='attr_advantage_devout_count' type='hidden')
+        div
+          | Devout
+          | #[span(name='attr_advantage_devout_count')]
```
Comment (jazeee) [inline ui/overviewTab/overviewTab.pug]: @johnlamping here is an example of what pug can do to replace line 8 - 57:

```
-
  var advantages = {
    'attr_advantage_healthy_count': 'Healthy',
    'attr_advantage_packmule_count': 'Pack Mule',
    'attr_advantage_strongarms_count': 'Strong Arms',
    'attr_advantage_brute_count': 'Brute',
    'attr_advantage_fast_count': 'Fast',
    'attr_advantage_athlete_count': 'Natural Athlete',
    'attr_advantage_alert_count': 'Alert',
    'attr_advantage_wary_count': 'Wary',
    'attr_advantage_mage_count': 'Mage',
    'attr_advantage_devout_count': 'Devout',
  }
```
and
```
      each advantageDescription, advantageName in advantages
        .conditional-something
          input(name=advantageName type='hidden')
          div
            | = advantageDescription
            | #[span(name=advantageName)]
```

I'll push a delta PR that shows this.

**Comment by johnlamping** [comment]
pug for the win!

File: ui/overviewTab/overviewTab.pug (hunk 5 above)
Comment (jazeee) [reply ui/overviewTab/overviewTab.pug]: I ended up changing to `beautify-js` to eliminate this pattern.

File: ui/overviewTab/overviewTab.pug (hunk 10 above)
Comment (jazeee) [reply ui/overviewTab/overviewTab.pug]: Merged into this PR. It works nicely now.

**Comment by jazeee** [comment]
@johnlamping I think we should go forward with pug.
There are some nice advantages from using it.

**Comment by johnlamping** [comment]
Sounds good to me!

**Comment by johnlamping** [comment]
Do you want approval now?

**Comment by jazeee** [comment]
Yeah, let's go forward with it. I think we can see some solid benefits.
It'll also identify and clear up html typos.
Also, can review the next pr too.

https://github.com/epic-power-rpg/roll20/pull/40

**Review by johnlamping — APPROVED** [review]

**Pull request merged**